### PR TITLE
Update CI to modern uv

### PR DESCRIPTION
**Pull request by manzt**
Saw a post on twitter... thought I'd take a stab.

Since I've seen you've embraced uv, I thought I'd suggest moving CI to the "new way" of running python stuff with uv.

This PR switches from manually managing viritual enviorments (e.g., the `uv pip`-oriented workflows) to the project workflows with the higher level `uv` commands. These APIs are new as of 0.3.0 and bring uv from a replacement for pip to something more like [cargo for python](https://astral.sh/blog/uv-unified-python-packaging) - while still being PEP compatible!

For example, in a fresh clone of the repo, you can run `uv run pytest` without any `pip` steps and uv will automatically 1.) make sure there is python iterpreter, 2.) create a virtual environment, 3.) install the dependencies, and 4.) run the command. uv performs aggressive caching, and will reuse the .venv if it can. This PR also checks in the uv lock file, which is managed by uv.

Also, `uv build` is a thing so I replaced that part of the workflow.

For Trusted Publishers, I'm not entirely sure what's going on. Reading the logs I saw the `write` permissions weren't set, but it seems like they are now.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -24,89 +19,54 @@ jobs:
...
       run: |
-        source .venv/bin/activate
-        ruff check .
+        uvx --python=3.12 ruff check .
 
     - name: Check formatting with black
       run: |
-        source .venv/bin/activate
-        black --check .
+        uvx --python=3.12 black --check .
```
Comment (manzt) [inline .github/workflows/ci.yml]: You might be interested in `uvx ruff format` and `uvx ruff format --check` for CI.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -24,89 +19,54 @@ jobs:
...
 
   publish:
     name: pypi-publish
     needs: [test, lint]
     runs-on: ubuntu-latest
-    environment:
-      name: release
-      url: https://pypi.org/p/grid_df
```
Comment (manzt) [inline .github/workflows/ci.yml:86]: Are these configured for the repo on PyPI?

Comment (manzt) [reply .github/workflows/ci.yml:86]: Here's a workflow of mine for a recent project: https://github.com/manzt/juv/blob/main/.github/workflows/release.yml

On PyPI, I just have the repo and workflow name configured (no enviroment):

image

**Comment by vsbuffalo** [comment]
Thank you so much @manzt! It still doesn't seem to be working but I really appreciate your PR and it was good to know the more modern uv approach. Here is my pypi configuration: 

Screenshot 2024-11-07 at 6 01 47 PM

I am still perplexed why this isn't working. I can push everything to pypi with `uv publish` but the GH actions are still failing for the same mysterious reason.

**Comment by manzt** [comment]
Super frustrating :/ .... I'm seriously at a loss of ideas for what could be going on.

**Pull request merged**